File: Minimum Path Sum.java
public class Solution {
    public int minPathSum(int[][] grid) {
        int M = grid.length;
        int N = grid[0].length;
        
        int[][] sum = new int[M][N];
        for (int i = 0; i < M; i++) {
            for (int j = 0; j < N; j++) {
                int left = Integer.MAX_VALUE;
                if (j > 0)
                    left = sum[i][j-1];
                int top = Integer.MAX_VALUE;
                if (i > 0)
                    top = sum[i-1][j];
                if (i == 0 && j == 0)
                    left = 0;
                sum[i][j] = Math.min(left, top) + grid[i][j];
            }
        }
        return sum[M-1][N-1];
    }
}

File: Minimum Path Sum.py
class Solution:
	# @param grid, a list of lists of integers
	# @return an integer
	def minPathSum(self, grid):
		BIG_VAL = 99999999
		M = len(grid)
		N = len(grid[0])
		sum = [[0]*N]*M
		for r in xrange(M):
			for c in xrange(N):
				left = BIG_VAL
				if (c > 0):
					left = sum[r][c-1]
				top = BIG_VAL
				if (r > 0):
					top = sum[r-1][c]
				if (r == 0 and c == 0):
					left = 0
				sum[r][c] = min(left, top) + grid[r][c]
		return sum[M-1][N-1]